pennylane/operation.py

```diff
@@ -542,6 +542,72 @@ def eigvals(self):
         """
         return self._eigvals(*self.parameters)
 
+    @staticmethod
+    def compute_terms(*params, **hyperparams):
+        r"""Static method to define the representation of this operation as a linear combination.
+
+        Each term in the linear combination is a pair of a scalar
+        value :math:`c_i` and an operator :math:`O_i`, so that the sum
+
+        .. math:: O = \sum_i c_i O_i
+
+        constructs this operator :math:`O`.
+
+        A ``NotImplementedError`` is raised if no representation by terms is defined.
+
+        .. note::
+            This method gets overwritten by subclasses to define the linear combination representation
+            of a particular operator.
+
+        Args:
+            params (list): trainable parameters of this operator, as stored in the ``parameters`` attribute
+            hyperparams (dict): non-trainable hyperparameters of this operator, as stored in the
+                ``hyperparameters["ops"]`` attribute
+
+        Returns:
+            list[tensor_like or float], list[.Operation]: list of coefficients and list of operations
+
+        **Example**
+
+        >>> qml.Hamiltonian().compute_terms([1., 2.], [qml.PauliX(0), qml.PauliZ(0)])
+        [1., 2.], [qml.PauliX(0), qml.PauliZ(0)]
+        """
+        return None
+
+    def terms(self):
+        r"""Representation of this operator as a linear combination.
+
+        Each term in the linear combination is a pair of a
+        scalar value :math:`c_i` and an operator :math:`O_i`, so that the sum
+
+        .. math:: O = \sum_i c_i O_i
+
+        constructs this operator :math:`O`.
+
+        .. note::
+            By default, this method calls the static method ``compute_terms``,
+            which is used by subclasses to define the concrete representation.
+
+        A ``NotImplementedError`` is raised if no representation through terms is defined.
+
+        Returns:
+            list[tensor_like or float], list[.Operation]: list of coefficients :math:`c_i`
+                and list of operations :math:`O_i`
+
+        **Example**
+
+        >>> qml.Hamiltonian([1., 2.], [qml.PauliX(0), qml.PauliZ(0)]).terms()
+        [1., 2.], [qml.PauliX(0), qml.PauliZ(0)]
+
+        The coefficients are differentiable and can be stored as tensors:
+
+        >>> import tensorflow as tf
+        >>> op = qml.Hamiltonian(tf.Variable([1., 2.]), [qml.PauliX(0), qml.PauliZ(0)])
+        >>> op.terms()[0]
+        [<tf.Tensor: shape=(), dtype=float32, numpy=1.0>, <tf.Tensor: shape=(), dtype=float32, numpy=2.0>]
```

Comment:
I'm curious about the unstacking here. Is this necessitated by the current logic? If so, is this temporary, and long term we want to always keep the tensor stacked here?

Reply:
Good question. At the moment the only reason to return something unstacked is that `self.parameters` is unstacked, and that is what terms gets...If you decide to store the full tensor (so that the hamiltonian has a single parameter), the stacked version would be returned. I think this is the plan, no?

Reply:
yes! it is